### PR TITLE
Fix ATen/OpenMP parallel_for behavior under mixed OpenMP runtimes on Windows

### LINKED ISSUE #176091: Windows CPU: SDPA FlashAttention gives wrong results with multiple intra-op threads (mixed OpenMP runtimes)
### 🐛 Describe the bug

On Windows, CPU `scaled_dot_product_attention` with `SDPBackend.FLASH_ATTENTION` gives wrong results when intra-op parallelism is enabled. Setting `torch.set_num_threads(1)` makes it correct again. I could not reproduce this on Linux (tested via Colab and WSL with the same shapes).

The break happens exactly when the CPU FlashAttention path goes parallel. `B=1, H=1, S=32` is fine, but `S=33` is already wrong. Bigger shapes are very wrong and sometimes non-deterministic.

### Repro : CPU SDPA FlashAttention wrong results

```python
import torch
import torch.nn.functional as F
from torch.nn.attention import SDPBackend, sdpa_kernel

def check(B, H, S, D=64):
    torch.manual_seed(42)
    q, k, v = [torch.randn(B, H, S, D) for _ in range(3)]
    with sdpa_kernel([SDPBackend.MATH]):
        ref = F.scaled_dot_product_attention(q, k, v).float()
    with sdpa_kernel([SDPBackend.FLASH_ATTENTION]):
        out = F.scaled_dot_product_attention(q, k, v).float()
    diff = (ref - out).abs().max().item()
    print(f"B={B} H={H} S={S} diff={diff} nan={torch.isnan(out).any().item()}")

torch.set_num_threads(4)
check(1, 1, 32)
check(1, 1, 33)
check(2, 4, 64)
torch.set_num_threads(1)
check(2, 4, 64)
````

Observed (Windows, 4 threads):

```
B=1 H=1 S=32 diff=5.96e-07 nan=False
B=1 H=1 S=33 diff=12.855   nan=False
B=2 H=4 S=64 diff=561.9  nan=False   (sometimes NaN, magnitude varies)
B=2 H=4 S=64 diff=6.56e-07 nan=False    (after set_num_threads(1))
```

I suspected a runtime/threading issue and checked what OpenMP DLLs are loaded. In a fresh process that only imports torch, I see both `vcomp140.dll` and `libomp140.x86_64.dll`.

```python
import torch, psutil, os
p = psutil.Process(os.getpid())
print("torch", torch.__version__)
for m in p.memory_maps():
    if any(x in m.path.upper() for x in ("VCOMP", "LIBOMP", "IOMP")):
        print(m.path)
```

Output:

```
C:\Windows\System32\libomp140.x86_64.dll
C:\Users\...\miniforge3\envs\pytorch-dev\vcomp140.dll
```

After more debugging i noticed that two failures are hapenning at high-level
1)  In the ATen OpenMP backend, `invoke_parallel` can end up with incorrect chunking and `at::get_thread_num() == 0` for all workers under mixed OpenMP runtimes. CPU FlashAttention indexes per-thread scratch by `at::get_thread_num()`, so all workers alias slot 0 and race, which matches the wrong/NaN behavior in Repro 1.

2.  `at::in_parallel_region()` may return false even while already executing inside an outer `at::parallel_for` under the same mixed-runtime condition. That can allow code to incorrectly start a nested `at::parallel_for` inside an existing parallel dispatch (e.g., via `parallel_dim_reduction`). On the fully-unpatched code this can be masked because (1) collapses chunking and both the outer and inner loops redundantly process the full range, but after fixing (1) the nested `parallel_for` can run with fewer threads than the code assumes, which can leave part of the output unwritten (partial writes in reductions).

### Expected behaviour 
diff < 1e-3 for all shapes, matching the MATH backend.

### Versions

PyTorch version: 2.12.0a0+gitcd2fbea
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Home (10.0.26100 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 4.2.3
Libc version: N/A

Python version: 3.11.14 | packaged by conda-forge | (main, Jan 26 2026, 23:39:55) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.26100-SP0
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090 Laptop GPU
Nvidia driver version: 581.95
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Name: Intel(R) Core(TM) i9-14900HX
Manufacturer: GenuineIntel
Family: 207
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 2200
MaxClockSpeed: 2200
L2CacheSize: 32768
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] numpy==2.4.2
[pip3] optree==0.18.0
[pip3] torch==2.12.0a0+gitcd2fbea
[conda] intel-openmp               2025.3.0               h57928b3_642           conda-forge
[conda] mkl                        2025.2.0               hac47afa_627           conda-forge
[conda] mkl-include                2025.3.0               h57928b3_455           conda-forge
[conda] numpy                      2.4.2                  pypi_0                 pypi
[conda] optree                     0.18.0                 pypi_0                 pypi
[conda] tbb                        2022.3.0               h3155e25_2             conda-forge
[conda] torch                      2.12.0a0+gitcd2fbea    pypi_0                 pypi

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @drisspg @liangel-02 @howardzhang-cv

### PR DESCRIPTION
**Pull request by gautamvarmadatla**
Fixes #176091

TLDR is On Windows, PyTorch loaded two OpenMP runtimes simultaneously, `vcomp140.dll` (MSVC, runs `#pragma omp parallel`) and `libiomp5md.dll` (libomp, resolves `omp_*()` API calls). These don't share state, so inside a VCOMP parallel region `omp_get_thread_num()` returns 0 for every thread and `omp_get_num_threads()` returns 1. `invoke_parallel` called both inside `#pragma omp parallel`, so all threads computed `tid=0` and a `chunk_size` equal to the full work range. CPU Flash Attention indexes per-thread scratch buffers by `at::get_thread_num()`, so all threads raced on slot 0 producing wrong results and NaN. `parallel_reduce` suffered the same, all threads wrote to `results[0]`. Separately, `omp_in_parallel()` returns false inside VCOMP regions, so `in_parallel_region()` was false on every worker and nested `parallel_for` calls would re-parallelize incorrectly.

### Changes

I fixed `invoke_parallel` in `ParallelOpenMP.h` by moving `num_threads` and `chunk_size` outside `#pragma omp parallel`, using `omp_get_max_threads()` instead of `omp_get_num_threads()`, and replacing `omp_get_thread_num()` with a `tid_counter` atomic so each thread gets a unique ID regardless of runtime. I also added a fallback loop on the calling thread to pick up any chunks VCOMP misses on cold start.

For `in_parallel_region()` I introduced `c10::ParallelPathGuard` in `c10/util/ParallelGuard.{h,cpp}`, a RAII thread-local set only in the parallel path of `parallel_for` and `parallel_reduce`. I also put it in `c10` rather than ATen because VCOMP creates its thread pool before `libtorch_cpu.dll` loads and `c10` thread-locals are reliably initialized by then. `in_parallel_region()` in `ParallelOpenMP.cpp` now checks `omp_in_parallel()` or `c10::ParallelPathGuard::is_active()`. finally i also wired the guard in `Parallel-inl.h` only on the parallel path, not the serial fallback, so `in_parallel_region()` stays false when `parallel_for` serializes.

And for testing, I added two tests to `test_parallel.cpp`, `ParallelForNoDuplicateWork` and `InParallelRegionDetectedInsideParallelFor`, and linked `test_parallel` with `caffe2::openmp` in `CMakeLists.txt` so the test binary actually compiles with the OMP flag and exercises real parallel paths.

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @nkhasbag-nv

### CODE REVIEW

**Comment by gautamvarmadatla** [comment]
hey @malfet mind taking a look when you get the chance :)

**Comment by malfet** [comment]
@gautamvarmadatla i understand what are you trying to achieve there, but wouldn't the better fix will be to link with the same runtime? Because there could be other issues
